### ISSUE TITLE
Rule Change: add option to ignore (stand-alone) class declarations with static initialization block in no-unused-vars

### ISSUE THREAD

**Issue by erosman** [eslint/eslint#17772]
### What rule do you want to change?

no-unused-vars

### What change do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
/* eslint no-unused-vars: "error" */

class ClassWithSIB {
  static {
    // …
  }
}
```

### What does the rule currently do for this code?

'ClassWithSIB' is defined but never used. ([no-unused-vars](https://eslint.org/docs/rules/no-unused-vars))

### What will the rule do after it's changed?

Ignore classes with [Static initialization blocks](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Static_initialization_blocks) in `no-unused-vars`

### Participation

- [ ] I am willing to submit a pull request to implement this change.

## Introduction

**SO-SIB-Class** (_definition for convenience_): A stand-alone `class` with a static initialization block that is not referenced elsewhere

- SO-SIB-Class is a valid OOP method
- Changing a SO-SIB-Class to anything else would be undesirable, and contrary to OOP, abstraction & encapsulation
- Currently, JavaScript `class` requires a name
  > Uncaught SyntaxError: class statement requires a name

## no-unused-vars

- It is my understanding that the purpose of `no-unused-vars` is to reduce superfluous variables
- A SO-SIB-Class is not a superfluous variable
- `no-unused-vars` **correctly** reports a SO-SIB-Class as an unused var
- Although correct in reporting, it necessitates multiple inline overrules which are in themselves superfluous 
- The aim is to introduce an option to eliminate the need for the superfluous inline overrules

### Previously

#17764, #17757

**Comment by mdjermanovic** [comment]
> **SO-SIB-Class** (_definition for convenience_): A stand-alone `class` with a static initialization block that is not referenced elsewhere
> 
> * SO-SIB-Class is a valid OOP method
> * Changing a SO-SIB-Class to anything else would be undesirable, and contrary to OOP, abstraction & encapsulation

Can you provide a link to a guideline that recommends this practice in JavaScript specifically? In programming languages where all code must be in classes this may be a common practice, but in JavaScript you can just write top-level code and wrap it in a block to isolate variables:

```js
{
    // …
}
```

**Comment by erosman** [comment]
The point is that the code is valid. 
You can look at the examples on [Static initialization blocks](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Static_initialization_blocks).

```js
class A {
  static field = "static field";
  static {
    console.log(this.field);
  }
}
```

> In programming languages where all code must be in classes this may be a common practice, but in JavaScript you can just write top-level code and wrap it in a block to isolate variables:

Should a programmer have to abandon a valid method of programming to suit the linter?

**Comment by mdjermanovic** [comment]
> You can look at the examples on [Static initialization blocks](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Static_initialization_blocks).
> ```js
> class A {
>   static field = "static field";
>   static {
>     console.log(this.field);
>   }
> }
> ```

It looks to me like a simplified example to show what `this` in static blocks refers to, rather than a recommendation to create classes just to wrap code in static blocks.

> > In programming languages where all code must be in classes this may be a common practice, but in JavaScript you can just write top-level code and wrap it in a block to isolate variables:
> 
> Should a programmer have to abandon a valid method of programming to suit the linter?

I personally think this is an unnecessary and uncommon use of classes in JavaScript. However, I can agree with the reasoning that this code can be valid (doing something it was intended to do) although the class name variable is unused, but is only necessary for the syntax to be valid, so I wouldn't be opposed to adding this option if other team members support this proposal.

**Comment by erosman** [comment]
> I personally think this is an unnecessary and uncommon use of classes in JavaScript.

It is probably uncommon, since the Static initialization blocks are fairly new.
_FF93 (2021-10-05), Ch94 (2021-09-21), ECMAScript 2022 (ES13)_

There are still many JavaScript programmers that write code with `var` and never use `class`.

As mentioned in another topic, it is a modern version of the standard static `class` that needs to auto-run some code.

**Standard class**
```js
class A {

  static run() {
    // do something
  }
}
A.run();
```

**ECMAScript 2022 class with SIB**
```js
class A {

  static {
    this.run();
  }

  static run() {
    // do something
  }
}
```

**Comment by erosman** [comment]
Let us also consider the issue from a different angle.

> [no-unused-vars](https://eslint.org/docs/latest/rules/no-unused-vars)
> 
> Variables that are declared and not used anywhere in the code are most likely an error due to incomplete refactoring. Such variables take up space in the code and can lead to confusion by readers.

The purpose of `no-unused-vars` is to identify variables that **have no relevance to the code** and **their removal have no consequences to the code execution**.

Classes with a [Static initialization blocks](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Static_initialization_blocks) **DO have relevance  to the code** and **their removal DO have consequences to the code execution**.

**Comment by Tanujkanti4441** [comment]
sorry! closed it accidentally.

**Comment by Rec0iL99** [comment]
ping @eslint/eslint-team

**Comment by JoshuaKGoldberg** [comment]
+1 to pushing back on this as a reasonable pattern to special-case for in JS code. If a class is not referenced anywhere else and is solely used for static purposes, there's no true benefit to methods on that class over a module-scoped function. This is covered in ESLint core by https://eslint.org/docs/latest/rules/class-methods-use-this.

...but, I'm not positive it's the `no-unused-vars` rule's realm of responsibility to handle this stylistic convention. Technically, erosman is right in https://github.com/eslint/eslint/issues/17772#issuecomment-1826125001: the value is technically used. It's just slightly obfuscated by the language how the value is used.

Personally, I think this issue is really two bugs:
* `class-methods-use-this` should report an error on a SO-SIB-Class, but doesn't
* `no-unused-var` should not report an error on a SO-SIB-Class, but does

Aside: there is a stylistic rule in typescript-eslint, https://typescript-eslint.io/rules/no-extraneous-class, that already flags this case.

**Comment by nzakas** [comment]
> * `class-methods-use-this` should report an error on a SO-SIB-Class, but doesn't

I'm not sure this is the purpose of this rule. The intent of the rule is to point out that some methods may be made static if they don't use `this`. In this case, we're talking about a static block becoming code outside of a class. If anything, that is likely a new rule.

I agree with @mdjermanovic that it makes sense to add a new option to `no-unused-vars` that would ignore classes with static inline blocks, which should be disabled by default. So, marking as accepted.

**Comment by erosman** [comment]
Great, thank you ...
When will it be available?

**Comment by Rec0iL99** [comment]
> When will it be available?

[`v9.0.0-beta.2`](https://github.com/eslint/eslint/releases/tag/v9.0.0-beta.2)

**Comment by erosman** [comment]
What about the v8?

**Comment by mdjermanovic** [comment]
This option will not be available in ESLint v8, only in v9.
